src/piton/extractors/omop_converter.py

```diff
@@ -224,57 +226,65 @@ def remove_pre_birth(patient: Patient) -> Optional[Patient]:
         return None
     else:
         return patient
 
 
-# One big flaw in our setup is that billing codes are assigned to the wrong date
-def move_billing_codes(patient: Patient) -> Patient:
-    new_events = []
-
-    for event in patient.events:
-        match event.event_type:
-            case _:
-                new_events.append(event)
-
-    new_events.sort(key=lambda a: a.start)
-
-    return Patient(patient_id=patient.patient_id, events=new_events)
+def _move_billing_codes(patient: Patient) -> Patient:
+    """Move billing codes to the end of each visit.
 
+    One issue with our OMOP extract is that billing codes are incorrectly assigned at the start of the visit.
+    This class fixes that by assigning them to the end of the visit.
+    """
 
-# Remove redunant rows
-def remove_redundant_rows(patient: Patient) -> Patient:
-    value_count: Dict[
-        Tuple[int, datetime.datetime],
-        Dict[Any, int],
-    ] = collections.defaultdict(lambda: collections.defaultdict(int))
+    end_visits: Dict[int, datetime.datetime] = {}
 
     for event in patient.events:
-        value_count[(event.code, event.start)][event.value] += 1
+        if event.event_type in ("lpch_pat_enc", "shc_pat_enc"):
+            if event.end is not None:
+                if event.visit_id is None:
+                    raise RuntimeError(
+                        f"Expected visit id for visit? {patient.patient_id} {event}"
+                    )
+                if end_visits.get(event.visit_id, event.end) != event.end:
+                    raise RuntimeError(
+                        f"Multiple end visits? {end_visits.get(event.visit_id)} {event}"
+                    )
+                end_visits[event.visit_id] = event.end
 
     new_events = []
 
     for event in patient.events:
-        counts = value_count[(event.code, event.start)]
-        if event.value is None:
-            # Remove if any of another type
-            if any(v is not None for v in counts.keys()):
-                continue
-
-        if counts[event.value] >= 2:
-            counts[event.value] -= 1
+        if event.event_type in ("lpch_pat_enc_dx", "shc_pat_enc_dx"):
+            if event.visit_id is None:
+                raise RuntimeError(
+                    f"Expected visit id for code {patient.patient_id} {event}"
+                )
+            end_visit = end_visits.get(event.visit_id)
+            if end_visit is None:
+                raise RuntimeError(
+                    f"Expected visit end for code {patient.patient_id} {event}"
+                )
+            new_events.append(dataclasses.replace(event, start=end_visit))
         else:
             new_events.append(event)
 
-    return Patient(patient.patient_id, new_events)
+    new_events.sort(key=lambda a: (a.start, a.code))
 
+    return Patient(patient_id=patient.patient_id, events=new_events)
+
+
+def _remove_nones(patient: Patient) -> Patient:
+    """Remove redundant NONE values.
```

Comment:
```suggestion
    """Remove duplicate codes within the same day if the duplicate code has a None value
```

Reply:
Are there any kinds of codes we can think of where the frequency of the code within a day matters even if the value for some of those codes are `None`?